### PR TITLE
Line types

### LINKED ISSUE #418: Add line types to internalize (#i:dim)
@cangiuli How about line types, which recognize `(abs ...)` as values but do not specify their end points? Path are subtypes of lines. (Better names are welcomed.)

### PR DESCRIPTION
**Pull request by jonsterling**
- [x] syntax
- [x] `hcom` dynamics
- [x] `coe` dynamics
- [x] rules

To resolve #418

### CODE REVIEW

File: src/redprl/refiner_types.fun [hunk 1]
```diff
@@ -1200,6 +1200,141 @@ struct
       end
   end
 
+  structure Line =
+  struct
+    val kindConstraintOnBase =
+      fn K.DISCRETE => K.DISCRETE
+       | K.KAN => K.KAN
+       | K.HCOM => K.HCOM
+       | K.COE => K.KAN
```
Comment (favonia) [inline src/redprl/refiner_types.fun]: `K.COE => K.COE`

File: src/redprl/refiner_types.fun [hunk 2]
```diff
@@ -1200,6 +1200,141 @@ struct
...
+        val Syn.LINE_TY (v, a1v) = Syn.out ty1
+        val ka = kindConstraintOnBase k
+
+        val w = alpha 0
+        val a0w = substVar (VarKit.toDim w, u) a0u
+        val a1w = substVar (VarKit.toDim w, v) a1v
+        val tyGoal = makeEqType (H @> (w, AJ.TERM O.DIM)) ((a0w, a1w), l, ka)
+
+        val a00 = substVar (Syn.into Syn.DIM0, u) a0u
+        val a01 = substVar (Syn.into Syn.DIM1, u) a0u
```
Comment (favonia) [inline src/redprl/refiner_types.fun]: We don't need `a00` and `a01`.

**Comment by favonia** [comment]
I wonder why GitHub says some check is still pending. From the travis web interface the commit seems to have passed everything.

**Comment by jonsterling** [comment]
@favonia When you get a chance, can you have another look at this and see if it looks correct? I have implemented my best guess at what the hcom and coe dynamics should be.

**Comment by jonsterling** [comment]
> Looks good. I might want to rename PATH_APP to DIM_APP and PATH_ABS to ABS.

OK! I don't have a strong opinion, and I think it's fine if you do that in some other PR.

Thank you for the review!

**Pull request merged**